drivers/nrf9160_gps/nrf9160_gps.c

```diff
@@ -0,0 +1,525 @@
+
+	int socket;
+
+	K_THREAD_STACK_MEMBER(thread_stack,
+			      CONFIG_NRF9160_GPS_THREAD_STACK_SIZE);
+	struct k_thread thread;
+	struct k_sem thread_run_sem;
+};
+
+static struct gps_data fresh_nmea;
+static struct gps_data fresh_pvt;
+
+static void copy_pvt(struct gps_pvt *dest, nrf_gnss_pvt_data_frame_t *src)
+{
+
+	dest->latitude = src->latitude;
+	dest->longitude = src->longitude;
+	dest->altitude = src->altitude;
+	dest->accuracy = src->accuracy;
+	dest->speed = src->speed;
+	dest->heading = src->heading;
+	dest->flags = src->flags;
+	dest->datetime.year = src->datetime.year;
+	dest->datetime.month = src->datetime.month;
+	dest->datetime.day = src->datetime.day;
+	dest->datetime.hour = src->datetime.hour;
+	dest->datetime.minute = src->datetime.minute;
+	dest->datetime.seconds = src->datetime.seconds;
+	dest->datetime.ms = src->datetime.ms;
+
+	for (size_t i = 0;
+	     i < MIN(NRF_GNSS_MAX_SATELLITES, GPS_MAX_SATELLITES); i++) {
+		dest->sv[i].sv = src->sv[i].sv;
+		dest->sv[i].cn0 = src->sv[i].cn0;
+		dest->sv[i].elevation = src->sv[i].elevation;
+		dest->sv[i].azimuth = src->sv[i].azimuth;
+		dest->sv[i].flags = src->sv[i].flags;
+		dest->sv[i].signal = src->sv[i].signal;
+	}
+}
+
+static bool is_fix(struct gps_pvt *pvt)
+{
+	return ((pvt->flags & NRF_GNSS_PVT_FLAG_FIX_VALID_BIT)
+		== NRF_GNSS_PVT_FLAG_FIX_VALID_BIT);
+}
+
+static u64_t fix_timestamp;
+
+static void print_satellite_stats(nrf_gnss_data_frame_t *pvt_data)
+{
+	u8_t  tracked          = 0;
+	u8_t  in_fix           = 0;
+	u8_t  unhealthy        = 0;
+
+	for (int i = 0; i < NRF_GNSS_MAX_SATELLITES; ++i) {
+
+		if ((pvt_data->pvt.sv[i].sv > 0) &&
+		    (pvt_data->pvt.sv[i].sv < 33)) {
+
+			tracked++;
+
+			if (pvt_data->pvt.sv[i].flags &
+					NRF_GNSS_PVT_FLAG_FIX_VALID_BIT) {
+				in_fix++;
+			}
+
+			if (pvt_data->pvt.sv[i].flags &
+					NRF_GNSS_SV_FLAG_UNHEALTHY) {
+				unhealthy++;
+			}
+		}
+	}
+
+	LOG_DBG("Tracking: %d Using: %d Unhealthy: %d", tracked,
+						       in_fix,
+						       unhealthy);
+
+	LOG_DBG("Seconds since last fix %lld",
+			(k_uptime_get() - fix_timestamp) / 1000);
+}
+
+static void gps_thread(int dev_ptr)
+{
+	struct device *dev = INT_TO_POINTER(dev_ptr);
+	struct gps_drv_data *drv_data = dev->driver_data;
+	int len;
+	nrf_gnss_data_frame_t raw_gps_data;
+	bool trigger_send = false;
+wait:
+	k_sem_take(&drv_data->thread_run_sem, K_FOREVER);
+
+	while (true) {
+		len = recv(drv_data->socket, &raw_gps_data,
+			       sizeof(nrf_gnss_data_frame_t), 0);
+		if (len <= 0) {
+			/* Is the GPS stopped, causing this error? */
+			if (!atomic_get(&drv_data->gps_is_active)) {
+				goto wait;
+			}
+			LOG_ERR("recv() returned error: %d", len);
+			continue;
+		}
+
+		print_satellite_stats(&raw_gps_data);
+
+		switch (raw_gps_data.data_id) {
+		case NRF_GNSS_PVT_DATA_ID:
+			copy_pvt(&fresh_pvt.pvt, &raw_gps_data.pvt);
+
+			if ((drv_data->trigger.chan == GPS_CHAN_PVT) &&
+			    (drv_data->trigger.type == GPS_TRIG_DATA_READY)) {
```

Comment:
As this module is today, it does not seem to be possible to set the trigger to _DATA_READY and _FIX, so you cannot get both in the application. Wouldn't it be better to send the event to the application regardless of which event is generated? Or expanding the API to accept more than one trigger?

Reply:
Indeed, this driver and API in general should be reworked, at least if it's still on the "work-in-progress state" where I left it before going on leave.
It's pretty much a copy of the sensor API, with the same limitations regarding triggers, and I've been awaiting the upstream rework of that API to see how GPS would better fit in there.

The plan was to have callback on multiple events, drop the mimicking of the sensor API and hopefully make use of the new upstream work. So if anyone wants to continue that work before I'm back, that would be great.

Reply:
What I did as a test was to butcher your implementation, and set the drv_data->trigger.type to the corresponding event, and call the handler() for each corresponding event. This way, I can filter in the application callback on the different events. The init function just omits the trigger type, so not nearly polished :)